src/custom/combinable_imports.rs

```diff
@@ -0,0 +1,103 @@
+use marker_api::{
+    ast::{AstPath, AstPathSegment},
+    prelude::*,
+    span::SpanSource,
+};
+
+marker_api::declare_lint! {
+    /// # What it does
+    /// TODO
+    COMBINABLE_IMPORTS,
+    Warn,
+}
+
+pub fn check_item<'ast>(cx: &'ast MarkerContext<'ast>, item: ast::ItemKind<'ast>) {
+    if let ast::ItemKind::Mod(module) = item
+        && !item.span().is_from_expansion()
+    {
+        check_items(cx, module.items());
+    }
+}
+
+pub fn check_crate<'ast>(cx: &'ast MarkerContext<'ast>, krate: ast::Crate<'ast>) {
+    check_items(cx, krate.items());
+}
+
+pub fn check_items<'ast>(cx: &'ast MarkerContext<'ast>, items: &[ast::ItemKind<'ast>]) {
+    let mut namespaces: Vec<UseItemInfo<'_>> = vec![];
+
+    for item in items {
+        // Only select imports
+        let import = if let ast::ItemKind::Use(import) = item
+            && let span = import.span()
+            && !span.is_from_expansion()
+        {
+            import
+        } else {
+            continue;
+        };
+
+        // Find merge candidate
+        let segs = stringify_path(import.use_path());
+        let import_info = UseItemInfo { segs, import };
+        let candidate = namespaces
+            .iter()
+            .fold((None, 0), |prev_selection, candidate| {
+                compare_candidate(&import_info, prev_selection, candidate)
+            });
+
+        // Emit lint or insert into `namespaces`
+        if let (Some(other), _) = candidate {
+            cx.emit_lint(COMBINABLE_IMPORTS, import, "this import can be merged...")
+                .span_note("...with this one", other.import.span());
+        }
+        namespaces.push(import_info);
+    }
+}
+
+fn stringify_path(path: &AstPath<'_>) -> Vec<String> {
+    path.segments()
+        .iter()
+        .map(|seg| seg.ident().name().to_string())
+        .collect()
+}
+
+fn compare_candidate<'a, 'ast>(
+    import: &UseItemInfo<'ast>,
+    best: (Option<&'a UseItemInfo<'ast>>, usize),
+    candidate: &'a UseItemInfo<'ast>,
+) -> (Option<&'a UseItemInfo<'ast>>, usize) {
+    let (best_candidate, best_count) = best;
+    let ctn = count_match(&import.segs, &candidate.segs);
+    if ctn > best_count
+        && !are_merged(
+            &import.import.use_path().segments()[ctn - 1],
+            &candidate.import.use_path().segments()[ctn - 1],
+        )
+    {
+        (Some(candidate), ctn)
+    } else {
+        (best_candidate, best_count)
+    }
+}
+
+fn count_match(a: &[String], b: &[String]) -> usize {
+    a.iter().zip(b.iter()).take_while(|(x, y)| x == y).count()
+}
+
+/// Checks if the two [`AstPathSegment`]s are the same or different.
```

Comment:
That's strange how `UseItem` works, because `use` is a tree, but it returns a single path.
As I see the use tree is represented as a number of use items that share the same prefix. That's quite unintuitive. There should definitely be something in the docs about that.

I created a PR https://github.com/rust-marker/marker/pull/282.
Maybe checking the identity of two `AstPathSegment` could also be part of their APIs?

Reply:
Yes it should be, `AstPathSemgments` will receive an ID which can be compared. They'll also implement the `EmissionNode` trait